### PR TITLE
Audio recording FrameProcessor

### PR DESCRIPTION
**Pull request by adriancowham**
Adds a `FrameProcessor` that will record a conversation between a bot and a human. The new `AudioBufferProcessor` is meant to be used a base class, it simply stores the conversation in a buffer. The intention is that developers can extend this class and do something interesting.

This PR also includes another `FrameProcessor`, `CanonicalMetrics`. This `FrameProcessor` is a concrete implementation of the `AudioBufferProcessor` and when the end of the call is reached, `CanonicalMetrics` will upload the audio to the Canonical AI platform for evaluation and metric collection.

This PR adds 2 examples in the `examples` directory demonstrating how to use both of the new `FrameProcessors`, they are in the `examples/chatbot-audio-recording` and `canonical-metrics` directories.

### CODE REVIEW

**Comment by kwindla** [comment]
@aconchillo, I talked to @adriancowham about this via email.

1. There have been a number of questions in Discord about how to build a pipeline element that can record a conversation. I think it's really useful to have a standard recording capability in Pipecat that doesn't depend on a transport implementation.
2. Canonical's new evals/observability platform for voice AI is really cool! Adrian's colleague Tom did a demo at the event last night in SF.

**Comment by aconchillo** [comment]
> @aconchillo, I talked to @adriancowham about this via email.
> 
> 1. There have been a number of questions in Discord about how to build a pipeline element that can record a conversation. I think it's really useful to have a standard recording capability in Pipecat that doesn't depend on a transport implementation.
> 2. Canonical's new evals/observability platform for voice AI is really cool! Adrian's colleague Tom did a demo at the event last night in SF.

This is fantastic. I'll review as soon as I can!

File: src/pipecat/processors/audio_buffer_processor.py
Comment (aconchillo) [inline src/pipecat/processors/audio_buffer_processor.py:28]: Please, use private fields (single underscore) so it's clear we don't expose this. 🙏

File: src/pipecat/processors/user_marker_processor.py
Comment (aconchillo) [inline src/pipecat/processors/user_marker_processor.py:10]: In theory this is not needed. All the audio coming from `transport.input()` is from the user. I believe what you have done in `AudioBufferProcessor` by checking `UserStartedSpeaking` and `UserStoppedSpeaking` should be enough.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (aconchillo) [inline src/pipecat/processors/canonical_metrics_processor.py:106]: Nit: add underscore to all private functions. 🙏

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (aconchillo) [inline src/pipecat/processors/canonical_metrics_processor.py:68]: Could we pass `api_key: str` to the constructor instead? This is the approach we take in many other services. Form outside you can do `os.environ.get("CANONICAL_API_KEY")` or get it from somewhere else.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (aconchillo) [inline src/pipecat/processors/canonical_metrics_processor.py:116]: If we follow my other comment here we would do `self._api_key`.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (aconchillo) [inline src/pipecat/processors/canonical_metrics_processor.py:111]: Same to `api_key`, we could probably add this to the constructor defaulting to `"https://voiceapp.canonical.chat/api/v1"`

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (aconchillo) [inline src/pipecat/processors/canonical_metrics_processor.py:137]: In many other services we pass `aiohttp_session` as an argument so we can provide an external session and we don't have to create it every time.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (aconchillo) [inline src/pipecat/processors/canonical_metrics_processor.py:144]: Here we could use `logger.error` and return so the bot doesn't stop working.

Comment (aconchillo) [reply src/pipecat/processors/canonical_metrics_processor.py:144]: Mmm... I actually think we capture the exception upstream, so maybe this is OK. I'll double-check

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (aconchillo) [inline src/pipecat/processors/canonical_metrics_processor.py:79]: We should not get anything after an `EndFrame` so we probably don't need `end_of_call`.

File: src/pipecat/processors/audio_buffer_processor.py
Comment (aconchillo) [inline src/pipecat/processors/audio_buffer_processor.py:63]: It seems this should be enough:

```
if self.user_audio:
```

File: src/pipecat/processors/audio_buffer_processor.py
Comment (aconchillo) [inline src/pipecat/processors/audio_buffer_processor.py:70]: Maybe:

```
if isinstance(frame, AudioRawFrame) and self.assistant_audio:
```

File: src/pipecat/processors/audio_buffer_processor.py
Comment (aconchillo) [reply src/pipecat/processors/audio_buffer_processor.py:28]: If we expose them,  we could use properties instead.

File: src/pipecat/processors/audio_buffer_processor.py
Comment (aconchillo) [inline src/pipecat/processors/audio_buffer_processor.py:1]: It'd be great if we could move this to `src/pipecat/processors/audio`.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (aconchillo) [inline src/pipecat/processors/canonical_metrics_processor.py:32]: I feel maybe this could be probably an `AIService` instead and we should then pass the `AudioBufferProcessor` as an argument.

So:

```
class CanonicalMetricsService(AIService):

    def __init__(self, audio_bufer_processor: AudioBufferProcessor, ...):
        ...
```

So `CanonicalMetricsServices` would live in `src/pipecat/services` with this approach. Another benefit is that you could then use any other `AudioBufferProcessor` implementation.

The only thing is that you then need to put both processors in the pipeline. But it's probably fine. So,

```
Pipeline([transport.input(), audio_buffer_processor, canonical, ....])
```

Another benefit is that other processors in the pipeline could use the `audio_buffer_processor`.

WDYT?

**Comment by adriancowham** [comment]
@aconchillo Thanks for the thoughtful feedback. I'll start working on this.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (adriancowham) [reply src/pipecat/processors/canonical_metrics_processor.py:32]: I like it. I'll make the change.

File: src/pipecat/processors/user_marker_processor.py
Comment (adriancowham) [reply src/pipecat/processors/user_marker_processor.py:10]: @aconchillo I can get this to work by only using `UserStartedSpeaking` and `UserStoppedSpeaking` but there is an longer than delay between when the assistant finishes and the human starts. This delay does not reflect what happened reality.  I'm pretty sure it's because audio from the user's mic is coming in and since there's no way to differentiate it, it gets added to the buffer while the assistant is talking. That is, the amount of extra delay added is the amount of time the assistant is speaking because all that silence from the user's mic is coming in and getting added to the buffer. There's a small mention of this in the discord, here: https://discord.com/channels/1239284677165056021/1278753324274814977/1279769874482204814

I'm happy to try an alternative implementation, but this is the best one I could come up with. The inflated delay is pretty bad, I don't think it would lead to usable feature.

Let me know your thoughts.

File: src/pipecat/processors/audio_buffer_processor.py
Comment (adriancowham) [reply src/pipecat/processors/audio_buffer_processor.py:63]: @aconchillo Gonna leave this one for now until we figure the latency issue described above.

File: src/pipecat/processors/audio_buffer_processor.py
Comment (adriancowham) [reply src/pipecat/processors/audio_buffer_processor.py:70]: @aconchillo Gonna leave this one for now until we figure the latency issue described above.

File: src/pipecat/processors/audio_buffer_processor.py
Comment (adriancowham) [reply src/pipecat/processors/audio_buffer_processor.py]: Done.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (adriancowham) [reply src/pipecat/processors/canonical_metrics_processor.py]: Done.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (adriancowham) [reply src/pipecat/processors/canonical_metrics_processor.py]: Done.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (adriancowham) [reply src/pipecat/processors/canonical_metrics_processor.py]: Done.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (adriancowham) [reply src/pipecat/processors/canonical_metrics_processor.py]: Done.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (adriancowham) [reply src/pipecat/processors/canonical_metrics_processor.py]: Done.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (adriancowham) [reply src/pipecat/processors/canonical_metrics_processor.py]: Done.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (adriancowham) [reply src/pipecat/processors/canonical_metrics_processor.py]: Done. I thought I saw a case where I was getting both of them.

File: src/pipecat/processors/audio_buffer_processor.py
Comment (adriancowham) [reply src/pipecat/processors/audio_buffer_processor.py]: Done.

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (adriancowham) [reply src/pipecat/processors/canonical_metrics_processor.py]: Done.

**Comment by adriancowham** [comment]
@aconchillo I pushed the changes from your feedback. I left you a comment, let me know what you think.

File: src/pipecat/processors/audio/audio_buffer_processor.py
Comment (aconchillo) [inline src/pipecat/processors/audio/audio_buffer_processor.py:30]: Use `loguru` logger if you need to log anything.  :pray:

File: src/pipecat/processors/canonical_metrics_processor.py
Comment (aconchillo) [reply src/pipecat/processors/canonical_metrics_processor.py]: Awesome! Thank you. Looking great!

**Comment by aconchillo** [comment]
> @aconchillo I pushed the changes from your feedback. I left you a comment, let me know what you think.

This looks great! I'll take a final look later today/tomorrow max.

File: src/pipecat/processors/user_marker_processor.py
Comment (aconchillo) [reply src/pipecat/processors/user_marker_processor.py:10]: So weird.... I'll take a look again.

File: src/pipecat/processors/audio/audio_buffer_processor.py
Comment (adriancowham) [reply src/pipecat/processors/audio/audio_buffer_processor.py]: My fault, I left a debug print statement in there. Fixed.

**Comment by adriancowham** [comment]
No rush.

File: src/pipecat/processors/user_marker_processor.py
Comment (aconchillo) [reply src/pipecat/processors/user_marker_processor.py:10]: Just for others who find this.... The reason this happens is because of this `DailyParams`:

```
                vad_audio_passthrough=True,
                vad_analyzer=SileroVADAnalyzer(),
                transcription_enabled=True,
```

The STT will happen on the Daily server so really no audio would be required in the pipeline. However, since we want to save the audio we have `vad_audio_passthrough` which continuously pushes audio frames.

**Comment by aconchillo** [comment]
This is the PR that I'll probably merge: https://github.com/pipecat-ai/pipecat/pull/480

You will no need the additional `UserMarkerProcessor` and you will need to check for `InputAudioRawFrame` instead.

**Comment by adriancowham** [comment]
Ok, sounds good. That's a massive PR, nice work on that! I'll keep my eye out for the merge and then update my PR.

**Comment by aconchillo** [comment]
> Ok, sounds good. That's a massive PR, nice work on that! I'll keep my eye out for the merge and then update my PR.

This is now merged. I'll take a final look once you merge. So excited to get this one in as well. Thank you for your patience!

**Comment by aconchillo** [comment]
Another massive PR went in to use Ruff as our formatter. Please, rebase and make sure the pipeline is green. I'll merge after that. Thank you! 🙏

**Comment by cyrilS-dev** [comment]
hey @aconchillo @adriancowham i did some test on my side, using `InputAudioRawFrame `& `OutputAudioRawFrame`
About the input : 
```python
async def process_frame(self, frame: Frame, direction: FrameDirection):
        await super().process_frame(frame, direction)
        
        if isinstance(frame, UserStartedSpeakingFrame):
            self._human_speaking = True
        elif isinstance(frame, UserStoppedSpeakingFrame):
            self._human_speaking= False
        elif isinstance(frame, InputAudioRawFrame) and self._human_speaking :
            self.audio_buffer.extend((frame.audio))
        
        await self.push_frame(frame, direction)
   ```
 is a wrong approach, because the VAD latency causes the beginning of the input audio to be missing for a few milliseconds.
 
 It means the input audio has to be recorded continuously, while the output audio is only recorded when the bot is speaking. 
 
 This creates a problem when combining the input and output audio, making the process more complicated.
 
 @aconchillo Do you have any insights? thanks

**Comment by aconchillo** [comment]
> hey @aconchillo @adriancowham i did some test on my side, using `InputAudioRawFrame `& `OutputAudioRawFrame` About the input :
> 
> ```python
> async def process_frame(self, frame: Frame, direction: FrameDirection):
>         await super().process_frame(frame, direction)
>         
>         if isinstance(frame, UserStartedSpeakingFrame):
>             self._human_speaking = True
>         elif isinstance(frame, UserStoppedSpeakingFrame):
>             self._human_speaking= False
>         elif isinstance(frame, InputAudioRawFrame) and self._human_speaking :
>             self.audio_buffer.extend((frame.audio))
>         
>         await self.push_frame(frame, direction)
> ```
> 
> is a wrong approach, because the VAD latency causes the beginning of the input audio to be missing for a few milliseconds.
> 
> It means the input audio has to be recorded continuously, while the output audio is only recorded when the bot is speaking.
> 
> This creates a problem when combining the input and output audio, making the process more complicated.
> 
> @aconchillo Do you have any insights? thanks

You are right, it would be a few milliseconds off. One solution is to keep, for example, a 1 second buffer. When you get the VAD event you just grab VADParams.start_secs from that buffer (maybe a bit more) and then start appending as we do now.

**Comment by aconchillo** [comment]
> 
> You are right, it would be a few milliseconds off. One solution is to keep, for example, a 1 second buffer. When you get the VAD event you just grab VADParams.start_secs from that buffer (maybe a bit more) and then start appending as we do now.

That said I’d rather merge this now (as soon as it’s rebased and reformatted) and then we can fix the issue later.

**Comment by cyrilS-dev** [comment]
> > hey @aconchillo @adriancowham i did some test on my side, using `InputAudioRawFrame `& `OutputAudioRawFrame` About the input :
> > ```python
> > async def process_frame(self, frame: Frame, direction: FrameDirection):
> >         await super().process_frame(frame, direction)
> >         
> >         if isinstance(frame, UserStartedSpeakingFrame):
> >             self._human_speaking = True
> >         elif isinstance(frame, UserStoppedSpeakingFrame):
> >             self._human_speaking= False
> >         elif isinstance(frame, InputAudioRawFrame) and self._human_speaking :
> >             self.audio_buffer.extend((frame.audio))
> >         
> >         await self.push_frame(frame, direction)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > is a wrong approach, because the VAD latency causes the beginning of the input audio to be missing for a few milliseconds.
> > It means the input audio has to be recorded continuously, while the output audio is only recorded when the bot is speaking.
> > This creates a problem when combining the input and output audio, making the process more complicated.
> > @aconchillo Do you have any insights? thanks
> 
> You are right, it would be a few milliseconds off. One solution is to keep, for example, a 1 second buffer. When you get the VAD event you just grab VADParams.start_secs from that buffer (maybe a bit more) and then start appending as we do now.

Thanks for the pointer, @aconchillo  ! I'm not sure I'll use the pipeline to record the audio since `InputAudioRawFrames` flow continuously through the pipeline, while `OutputAudioRawFrames` are only present when the TTS is speaking. This means that silences would need to be inserted into the output audio to maintain sync with the input. Using LiveKit Egress seems like a more precise approach.

**Comment by aconchillo** [comment]
> > > hey @aconchillo @adriancowham i did some test on my side, using `InputAudioRawFrame `& `OutputAudioRawFrame` About the input :
> > > ```python
> > > async def process_frame(self, frame: Frame, direction: FrameDirection):
> > >         await super().process_frame(frame, direction)
> > >         
> > >         if isinstance(frame, UserStartedSpeakingFrame):
> > >             self._human_speaking = True
> > >         elif isinstance(frame, UserStoppedSpeakingFrame):
> > >             self._human_speaking= False
> > >         elif isinstance(frame, InputAudioRawFrame) and self._human_speaking :
> > >             self.audio_buffer.extend((frame.audio))
> > >         
> > >         await self.push_frame(frame, direction)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > is a wrong approach, because the VAD latency causes the beginning of the input audio to be missing for a few milliseconds.
> > > It means the input audio has to be recorded continuously, while the output audio is only recorded when the bot is speaking.
> > > This creates a problem when combining the input and output audio, making the process more complicated.
> > > @aconchillo Do you have any insights? thanks
> > 
> > 
> > You are right, it would be a few milliseconds off. One solution is to keep, for example, a 1 second buffer. When you get the VAD event you just grab VADParams.start_secs from that buffer (maybe a bit more) and then start appending as we do now.
> 
> Thanks for the pointer, @aconchillo ! I'm not sure I'll use the pipeline to record the audio since `InputAudioRawFrames` flow continuously through the pipeline, while `OutputAudioRawFrames` are only present when the TTS is speaking. This means that silences would need to be inserted into the output audio to maintain sync with the input. Even with the idea of recording the input audio based on VAD events, i would still need to reconstruct the silences by using timestamps on the frames to maintain proper alignment and sync. Using LiveKit Egress seems like a more precise and simpler approach.

That's sounds reasonable. We have recording support at Daily as well and if you use a service transport it makes sense to use the provided recording.

**Comment by adriancowham** [comment]
@aconchillo @cyrilS-dev I just got back from out of town and I'm all caught up. I plan to jump on this tomorrow. I'll play around with VAD stuff while I'm in there. Thanks for the insights!

**Comment by adriancowham** [comment]
@aconchillo I nuked `UserMarkerProcessor` and now using `InputAudioRawFrame`. I also changed audio processing code to save to stereo. Things work well, however, when the bot is interrupted the recording has the bot's full response but the live playback doesn't. I'm not sure how to address, any guidance would be appreciated. Thanks!

**Comment by adriancowham** [comment]
@aconchillo Wanted to get this back on your radar, no rush, but curious to hear your thoughts on how best to handle interruptions. Thanks!

**Comment by aconchillo** [comment]
> @aconchillo Wanted to get this back on your radar, no rush, but curious to hear your thoughts on how best to handle interruptions. Thanks!

Sorry for the delay. I've been heads down on something else. 

I believe all you have to do is put the audio processor after `transport.output()`.

So, this:
```
            transport.output(),
            audiobuffer,  # used to buffer the audio in the pipeline
```

instead of

```
            audiobuffer,  # used to buffer the audio in the pipeline
            transport.output(),
```

**Comment by adriancowham** [comment]
@aconchillo Ah, makes perfect sense, I'll give it a go. Thanks!

**Comment by adriancowham** [comment]
@aconchillo It's a lot better now, thanks for the tip. Hopefully this PR is ready, let me know if there's anything else I can do?

Thank you!

File: src/pipecat/processors/audio/audio_buffer_processor.py
Comment (aconchillo) [inline src/pipecat/processors/audio/audio_buffer_processor.py:17]: pass `**kwargs`

File: src/pipecat/processors/audio/audio_buffer_processor.py
Comment (aconchillo) [inline src/pipecat/processors/audio/audio_buffer_processor.py:31]: pass `**kwargs`

File: src/pipecat/processors/audio/audio_buffer_processor.py
Comment (aconchillo) [inline src/pipecat/processors/audio/audio_buffer_processor.py:58]: this should always be 2 because every sample is a signed 16-bit integer. If we change sample rate to 44100 this would break.

File: src/pipecat/processors/audio/audio_buffer_processor.py
Comment (aconchillo) [inline src/pipecat/processors/audio/audio_buffer_processor.py:85]: nit: remove outter parenthesis

File: src/pipecat/processors/audio/audio_buffer_processor.py
Comment (aconchillo) [inline src/pipecat/processors/audio/audio_buffer_processor.py:98]: nit: remove outter parenthesis

File: src/pipecat/processors/audio/audio_buffer_processor.py
Comment (adriancowham) [reply src/pipecat/processors/audio/audio_buffer_processor.py:29]: Ha, no. Good catch.

File: src/pipecat/services/canonical.py
Comment (aconchillo) [inline src/pipecat/services/canonical.py]: force keyword-only arguments with

```
self, *, ...
```

File: src/pipecat/services/canonical.py
Comment (aconchillo) [inline src/pipecat/services/canonical.py:69]: pass `**kwargs`

**Comment by aconchillo** [comment]
> @aconchillo It's a lot better now, thanks for the tip. Hopefully this PR is ready, let me know if there's anything else I can do?
> 
> Thank you!

Added some last minor comments. All looks great. I'll merge after those.

**Comment by adriancowham** [comment]
No problem at all, I appreciate all the feedback and the time you've put into this PR.

**Comment by aconchillo** [comment]
LGTM! Thank you so much for all the work!!!

**Review by aconchillo — APPROVED** [review]

**Pull request merged**
